DerivativeSignal update so GPU_ACTIVITY reads 0 instead of NAN.

asmaalrawi opened the pull request:
- Relates to #3550 

Update DerivativeSignal to add a new constructor including m_nan_replace, which replaces NANs due to the denominator evaluating to 0. The default behavior is to output NAN (e.g. oversampling).  For GPU utilization/activity signals, the timer does not update if there's no activity on the card, so 0/0 should evaluate to 0.

In the future, the aggregator should also be updated to handle "NAN" appropriately (rather than propagating NAN).

asmaalrawi committed:
DerivativeSignal update so GPU_ACTIVITY reads 0 instead of NAN.

Commit:
Reordered DerivativeSignal member variables.

Commit:
Passed m_nan_replace into static function.

Commit:
Cleaned up DerivativeSignalTest

Commit:
Fixed DerivativeSignalTest read_nan to expect 0.

dannosliwcd commented on libgeopmd/src/DerivativeSignal.cpp:
```diff
@@ -26,6 +26,7 @@ namespace geopm
         , m_is_batch_ready(false)
         , m_sleep_time(sleep_time)
         , m_last_result(NAN)
+        , m_nan_replace(0)
```
Since this is intended for only 2 signals, should the default behavior match the current behavior? Then we could just apply the zero-fill for the two signals where we know that's desirable.

asmaalrawi replied:
Other signals are currently using MSR::TIME which is guaranteed to monotonically increase with each sample. When new signals are introduced that use timers that might not increase with each read, we will add a new constructor that allows m_nan_replace to be declared.

dannosliwcd reviewed (commented):
Are there any cases resulting from this change where GEOPM traces would contain false data that otherwise would have been reported as indeterminate data that users could find with a nan query?

Can you check how this behaves on traces with `LEVELZERO::GPU_POWER` and `LEVELZERO::GPU_CORE_POWER`? If neighboring GEOPM samples are too close together (I'm not sure whether LevelZero explicitly defines an expected energy-counter update period), then the GPUs will occasionally show 0 W. Maybe that problem only happens if the trace-head or any later 8 (window width selected by LevelZeroIOGroup) samples in a row are too close, so that's less likely.

Commit:
Added is_sampled() method to Sample class

- to enable derivative signals to track state
- to calculate derivative only when signal is updated

Co-authored-by: Christopher M Cantalupo <christopher.m.cantalupo@intel.com>
Signed-off-by: Asma H. Al-Rawi <asma.al-rawi@intel.com>

Commit:
Added DerivativeSignal test for oversampling.

Commit:
Added DerivativeSignal constructor specifying m_nan_replace.

- Changed LevelZeroIOGroup derivative signals to specify desired nan
behavior when oversampling
- Updated DerivativeSignalTest with new constructor

Signed-off-by: Asma H. Al-Rawi <asma.al-rawi@intel.com>

asmaalrawi commented:
Changed this so that default behavior when oversampling is "NAN" and added a constructor that allows each DerivativeSignal to choose a default value to replace any samples where the timestamp hasn't changed.

I tested with GPU_POWER and GPU_CORE_POWER using geopmsession and was not able to sample quickly enough to get a NAN or a 0 after the first sample.

cmcantalupo commented on libgeopmd/src/DerivativeSignal.cpp:
```diff
@@ -26,7 +26,25 @@ namespace geopm
+        , m_is_batch_ready(false)
+        , m_sleep_time(sleep_time)
+        , m_last_result(NAN)
+        , m_nan_replace(nan_replace)
```
Avoid the copy/pasted code to maintain by calling the nan_replace constructor in the one with out nan_replace

cmcantalupo committed:
Remove copy-pasted code

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
Note this does change the behavior for the theoretical case where a different signal's denominator is not updated.  Previous to this change, the result would be the last calculated value, now it will be NaN

Commit:
Fixed new sample comment in DerivativeSignal.

bgeltz commented:
An issue should still be filed around having the reporter / aggregator ignore NaNs.  As it is, if a NaN is aggregated, the reporter will effectively ignore that signal forever.

asmaalrawi commented:
> An issue should still be filed around having the reporter / aggregator ignore NaNs. As it is, if a NaN is aggregated, the reporter will effectively ignore that signal forever.

https://github.com/geopm/geopm/issues/3592
